crates/oxc_query/src/edges.rs

```diff
@@ -0,0 +1,2170 @@
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::ancestors(contexts, adapter)
+    }
+
+    pub(super) fn parent<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+        adapter: &'a Adapter<'b>,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::parents(contexts, adapter)
+    }
+
+    pub(super) fn span<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::get_span(contexts)
+    }
+}
+
+pub(super) fn resolve_assignment_type_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "span" => assignment_type::span(contexts, resolve_info),
+        _ => {
+            unreachable!(
+                "attempted to resolve unexpected edge '{edge_name}' on type 'AssignmentType'"
+            )
+        }
+    }
+}
+
+mod assignment_type {
+    use trustfall::provider::{
+        ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo, VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+
+    pub(super) fn span<'a, 'b: 'a>(
+        contexts: ContextIterator<'a, Vertex<'b>>,
+        _resolve_info: &ResolveEdgeInfo,
+    ) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+        super::get_span(contexts)
+    }
+}
+
+pub(super) fn resolve_class_edge<'a, 'b: 'a>(
+    contexts: ContextIterator<'a, Vertex<'b>>,
+    edge_name: &str,
+    _parameters: &EdgeParameters,
+    resolve_info: &ResolveEdgeInfo,
+) -> ContextOutcomeIterator<'a, Vertex<'b>, VertexIterator<'a, Vertex<'b>>> {
+    match edge_name {
+        "constructor" => class::constructor(contexts, resolve_info),
+        "entire_class_span" => class::entire_class_span(contexts, resolve_info),
+        "getter" => class::getter(contexts, resolve_info),
+        "method" => class::method(contexts, resolve_info),
+        "name_span" => class::name_span(contexts, resolve_info),
+        "property" => class::property(contexts, resolve_info),
+        "setter" => class::setter(contexts, resolve_info),
+        _ => {
+            unreachable!("attempted to resolve unexpected edge '{edge_name}' on type 'Class'")
+        }
+    }
+}
+
+mod class {
+    use std::rc::Rc;
+
+    use oxc_ast::ast::{ClassElement, MethodDefinitionKind};
+    use trustfall::provider::{
+        resolve_neighbors_with, ContextIterator, ContextOutcomeIterator, ResolveEdgeInfo,
+        VertexIterator,
+    };
+
+    use super::super::vertex::Vertex;
+    use crate::vertex::{ClassMethodVertex, ClassPropertyVertex};
+
+    macro_rules! class_fn_edge_implem {
```

Comment:
is ti confusing to use a macro like this? It's not super much code but when we make code changes I want it to be clear that every time this macro is used, every macro invocation's code should change too.

Reply:
the macro is mostly okay, it's a tradeoff either way